core/src/main/java/org/elasticsearch/ingest/PipelineStore.java

```diff
@@ -183,270 +182,98 @@ public void put(PutPipelineRequest request, ActionListener<IndexResponse> listen
-                    innerPut(request, listener);
-                }
+            @Override
+            protected PutPipelineResponse newResponse(boolean acknowledged) {
+                return new PutPipelineResponse(acknowledged);
+            }
 
-                @Override
-                public void onFailure(Throwable e) {
-                    listener.onFailure(e);
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                IngestMetadata currentIngestMetadata = currentState.metaData().custom(IngestMetadata.TYPE);
+                Map<String, Map<String, Object>> pipelines;
+                if (currentIngestMetadata != null) {
+                    pipelines = new HashMap<>(currentIngestMetadata.getPipelines());
+                } else {
+                    pipelines = new HashMap<>();
                 }
-            });
-        }
-    }
-
-    private void innerPut(PutPipelineRequest request, ActionListener<IndexResponse> listener) {
-        IndexRequest indexRequest = new IndexRequest(request);
-        indexRequest.index(PipelineStore.INDEX);
-        indexRequest.type(PipelineStore.TYPE);
-        indexRequest.id(request.id());
-        indexRequest.source(request.source());
-        indexRequest.refresh(true);
-        client.index(indexRequest, handleWriteResponseAndReloadPipelines(listener));
+                Map<String, Object> pipelineConfig = XContentHelper.convertToMap(request.source(), false).v2();
+                pipelines.put(request.id(), pipelineConfig);
+
+                ClusterState.Builder newState = ClusterState.builder(currentState);
+                newState.metaData(MetaData.builder(currentState.getMetaData())
+                    .putCustom(IngestMetadata.TYPE, new IngestMetadata(pipelines))
+                    .build());
+                return newState.build();
+            }
+        });
     }
 
     /**
      * Returns the pipeline by the specified id
      */
     public Pipeline get(String id) {
-        ensureReady();
-
-        PipelineDefinition ref = pipelines.get(id);
-        if (ref != null) {
-            return ref.getPipeline();
-        } else {
-            return null;
-        }
+        return pipelines.get(id);
     }
 
     public Map<String, Processor.Factory> getProcessorFactoryRegistry() {
         return processorFactoryRegistry;
     }
 
-    public List<PipelineDefinition> getReference(String... ids) {
-        ensureReady();
-
-        List<PipelineDefinition> result = new ArrayList<>(ids.length);
+    /**
+     * @return a pipeline as maps of maps by id. Multiple pipelines will be returned if multiple ids are provided.
+     */
+    // Returning map of maps instead of pipeline, because pipeline / processor interface doesn't
+    // know how to serialize itself.
+    public Map<String, Map<String, Object>> getPipelines(String... ids) {
```

Comment:
if we add PipelineConfiguration as suggested above, it could be returned by the get pipeline api and things would be a bit nicer?